doc/designs/quic-design/quic-fifm.md

```diff
@@ -0,0 +1,531 @@
+allocation made per transmitted packet. The TX packetiser will obtain
+a `QUIC_TXPIM_PKT` structure from the TXPIM, fill in the structure including
+the ACK Manager data, and submit it via the FIFD which we introduce below.
+
+The TXPIM does do not anything with the `QUIC_TXPIM_PKT` structure itself other
+than managing its allocation and manipulation. Constructive use of the data kept
+in the TXPIM is made by the FIFD.
+
+### API
+
+```c
+/*
+ * QUIC Transmitted Packet Information Manager
+ * ===========================================
+ */
+typedef struct quic_txpim_st QUIC_TXPIM;
+
+typedef struct quic_txpim_pkt_st {
+    /* ACKM-specific data. Caller should fill this. */
+    QUIC_ACKM_TX_PKT    ackm_pkt;
+
+    /* Linked list of CFQ items in this packet. */
+    QUIC_CFQ_ITEM      *retx_head;
+
+    /* Reserved for FIFD use. */
+    QUIC_FIFD          *fifd;
+
+    /* Regenerate-strategy frames. */
+    unsigned int        had_handshake_done          : 1;
+    unsigned int        had_max_data_frame          : 1;
+    unsigned int        had_max_streams_bidi_frame  : 1;
+    unsigned int        had_max_streams_uni_frame   : 1;
+    unsigned int        had_ack_frame               : 1;
+
+    /* Private data follows. */
+} QUIC_TXPIM_PKT;
+
+/* Represents a range of bytes in an application or CRYPTO stream. */
+typedef struct quic_txpim_chunk_st {
+    /* The stream ID, or UINT64_MAX for the CRYPTO stream. */
+    uint64_t        stream_id;
+    /*
+     * The inclusive range of bytes in the stream. Exceptionally, if end <
+     * start, designates a frame of zero length (used for FIN-only frames).
+     */
+    uint64_t        start, end;
+    /*
+     * Whether a FIN was sent for this stream in the packet. Not valid for
+     * CRYPTO stream.
+     */
+    unsigned int    has_fin : 1;
+} QUIC_TXPIM_CHUNK;
+
+QUIC_TXPIM *ossl_quic_txpim_new(void);
+void ossl_quic_txpim_free(QUIC_TXPIM *txpim);
+
+/*
+ * Allocates a new QUIC_TXPIM_PKT structure from the pool. Returns NULL on
+ * failure. The returned structure is cleared of all data and is in a fresh
+ * initial state.
+ */
+QUIC_TXPIM_PKT *ossl_quic_txpim_pkt_alloc(QUIC_TXPIM *txpim);
+
+/*
+ * Releases the TXPIM packet, returning it to the pool.
+ */
+void ossl_quic_txpim_pkt_release(QUIC_TXPIM *txpim, QUIC_TXPIM_PKT *fpkt);
+
+/* Clears the chunk list of the packet, removing all entries. */
+void ossl_quic_txpim_pkt_clear_chunks(QUIC_TXPIM_PKT *fpkt);
+
+/* Appends a chunk to the packet. The structure is copied. */
+int ossl_quic_txpim_pkt_append_chunk(QUIC_TXPIM_PKT *fpkt,
+                                    const QUIC_TXPIM_CHUNK *chunk);
+
+/* Adds a CFQ item to the packet by prepending it to the retx_head list. */
+void ossl_quic_txpim_pkt_add_cfq_item(QUIC_TXPIM_PKT *fpkt,
+                                      QUIC_CFQ_ITEM *item);
+
+/*
+ * Returns a pointer to an array of stream chunk information structures for the
```

Comment:
This approach was chosen over using OPENSSL_STACK because OPENSSL_STACK only stores a `void *`. This design avoids having one heap allocation per chunk.

Reply:
> because OPENSSL_STACK only stores a void *

Err, no? You're not supposed to use OPENSSL_STACK directly. You use the wrapper macros `DEFINE_STACK_OF`, `STACK_OF` etc. This provides type specific stacks.

Reply:
Safestack is a set of macros on top of OPENSSL_STACK, which stores an array of `void *`. So it needs to be used with objects allocated on the heap:
```c
    static ossl_unused ossl_inline int sk_##t1##_push(STACK_OF(t1) *sk, t2 *ptr) \
    { \
        return OPENSSL_sk_push((OPENSSL_STACK *)sk, (const void *)ptr); \
    } \
```

The implementation here uses a resizeable array of inline structures, i.e. can store n chunks in a single heap allocation.

Reply:
It doesn't _need_ to be used with allocated objects.  Using it with allocated objects makes freeing easier.  Management of the memory allocation could be done in a number of way: list of allocated chunks, a boolean field indicating the first in a block or whatever.

In case you hadn't noticed, I've very much against re-inventing things when an adequate solution already exists :wink: